src/js/components/SchemaField.js

```diff
@@ -0,0 +1,315 @@
+    } = props;
+
+    let field = (
+      <FieldInput
+        id={name}
+        type={"text"}
+        className={"field-input-text-narrow"}
+        autoFocus={autofocus}
+        name={name}
+        value={formData}
+        onChange={event => onChange(event.target.value)}
+        onBlur={onBlur && (event => onBlur(name, event.target.value))}
+        onFocus={onFocus && (event => onFocus(name, event.target.value))}
+      />
+    );
+    if (autofocus) {
+      field = (
+        <FieldAutofocus>
+          {field}
+        </FieldAutofocus>
+      );
+    }
+
+    return (
+      <div>
+        <FieldLabel>
+          {this.getFieldHeading(required, name, schema.description)}
+        </FieldLabel>
+        {field}
+        <FieldError>{errorMessage}</FieldError>
+      </div>
+    );
+  }
+
+  renderNumberInput(errorMessage, autofocus, props) {
+    const {
+      required,
+      name,
+      schema,
+      formData,
+      onChange,
+      onBlur,
+      onFocus
+    } = props;
+
+    const handleChange = event => {
+      const value = event.target.value;
+
+      if (value === "") {
+        onChange("");
+      } else {
+        onChange(Number(value));
+      }
+    };
+
+    let field = (
+      <FieldInput
+        id={name}
+        type={"number"}
+        className={"field-input-number-narrow"}
+        autoFocus={autofocus}
+        name={name}
+        value={formData}
+        onChange={handleChange.bind(this)}
+        onBlur={onBlur && (event => onBlur(name, event.target.value))}
+        onFocus={onFocus && (event => onFocus(name, event.target.value))}
+      />
+    );
+    if (autofocus) {
+      field = (
+        <FieldAutofocus>
+          {field}
+        </FieldAutofocus>
+      );
+    }
+
+    return (
+      <div>
+        <FieldLabel>
+          {this.getFieldHeading(required, name, schema.description)}
+        </FieldLabel>
+        {field}
+        <FieldError>{errorMessage}</FieldError>
+      </div>
+    );
+  }
+
+  getFieldHeading(required, name = "", description) {
+    let requiredSymbol = null;
+    if (required) {
+      requiredSymbol = (
+        <FormGroupHeadingContent className="text-danger" primary={false}>
+          *
+        </FormGroupHeadingContent>
+      );
+    }
+
+    return (
+      <FormGroupHeading>
+        <FormGroupHeadingContent primary={true}>
+          {name.split(/_|-/).join(" ")}
+        </FormGroupHeadingContent>
+        {requiredSymbol}
+        <FormGroupHeadingContent primary={false}>
+          <Tooltip
+            content={description}
+            interactive={true}
+            maxWidth={300}
+            wrapText={true}
+          >
+            <Icon color="grey" id="circle-question" size="mini" />
+          </Tooltip>
+        </FormGroupHeadingContent>
+      </FormGroupHeading>
+    );
+  }
+
+  getFieldContent(errorMessage, autofocus) {
+    const { schema } = this.props;
+    const RADIO_SELECT_THRESHOLD = 4;
+
+    switch (schema.type) {
```

Comment:
🤔 wonder if we should add a `default` case for unknown fields...

Reply:
I think this a good idea. I've added a default that renders text input. Text input seems least likely to explode with a strange schema type:

```
default:
    return this.renderTextInput(errorMessage, autofocus, this.props);
```